binary_transparency/firmware/internal/crypto/signature.go

```diff
@@ -0,0 +1,130 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crypto
+
+import (
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha512"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/golang/glog"
+)
+
+func getPrivateKey() (*rsa.PrivateKey, error) {
+	mKey := strings.NewReader(RSApri)
+	priv, err := ioutil.ReadAll(mKey)
+	if err != nil {
+		return nil, fmt.Errorf("Read failed! %s", err)
+	}
+
+	privPem, _ := pem.Decode(priv)
+	if privPem.Type != "RSA PRIVATE KEY" {
+		return nil, fmt.Errorf("RSA private key is of the wrong type %s", privPem.Type)
+	}
+
+	var parsedKey interface{}
+	if parsedKey, err = x509.ParsePKCS1PrivateKey(privPem.Bytes); err != nil {
+		return nil, fmt.Errorf("Unable to parse RSA private key %v", err)
+	}
+
+	privateKey, ok := parsedKey.(*rsa.PrivateKey)
+	if !ok {
+		return nil, fmt.Errorf("Unable to parse RSA private key %v", err)
+	}
+	return privateKey, nil
+}
+
+func getPublicKey() (*rsa.PublicKey, error) {
+	mKey := strings.NewReader(RSApub)
+	pub, err := ioutil.ReadAll(mKey)
+	if err != nil {
+		glog.Exitf("Pubkey read failed! %s", err)
+	}
+
+	pubPem, _ := pem.Decode(pub)
+	if pubPem == nil {
+		return nil, fmt.Errorf("pem decoded to nil")
+	}
+	if pubPem.Type != "RSA PUBLIC KEY" {
+		return nil, fmt.Errorf("RSA public key is of the wrong type %s", pubPem.Type)
+	}
+
+	var parsedKey interface{}
+	if parsedKey, err = x509.ParsePKCS1PublicKey(pubPem.Bytes); err != nil {
+		return nil, fmt.Errorf("Unable to parse RSA public key, generating a temp one %v", err)
+	}
+
+	var pubKey *rsa.PublicKey
+	var ok bool
+	if pubKey, ok = parsedKey.(*rsa.PublicKey); !ok {
+		return nil, fmt.Errorf("Unable to parse RSA public key")
+	}
+
+	return pubKey, nil
+}
+
+//SignMessage is used to sign the Statement
+func SignMessage(msg []byte) (sig []byte, err error) {
+	// Before signing, we need to hash the message
+	// The hash is what we actually sign
+	msgHash := sha512.New()
+	if _, err = msgHash.Write(msg); err != nil {
+		glog.Exitf("Message hashing failed %v", err)
+	}
+	msgHashSum := msgHash.Sum(nil)
+
+	// Get the required key for signing
+	key, err := getPrivateKey()
+	if err != nil {
+		glog.Exitf("Private key fetch failed %v", err)
+	}
+
+	signature, err := rsa.SignPSS(rand.Reader, key, crypto.SHA512, msgHashSum, nil)
+	if err != nil {
+		glog.Exitf("Signing failed for firmware statement: %v", err)
+		return nil, err
+	}
+	return signature, nil
+}
+
+//VerifySignature is used to verify the incoming message
+func VerifySignature(msg []byte, signature []byte) (verified bool, err error) {
```

Comment:
I think this only really needs to return an `error` - if it's `nil` then the signature verified, otherwise there was a problem (either something went wrong leading up to the verification, or the verification itself failed - note that `rsa.Verify...()` returns just an `error`).

Reply:
Done, now it just returns error